File: compare-version-numbers/solution165.py
# https://leetcode.com/problems/compare-version-numbers/

class Solution165:

    # T: O(max(n,m)), S: O(n+m)
    def compareVersion(self, version1: str, version2: str) -> int:
        v1_parts = version1.split(".")
        v2_parts = version2.split(".")

        max_length = max(len(v1_parts), len(v2_parts))

        for i in range(max_length):
            v1_val = int(v1_parts[i]) if i < len(v1_parts) else 0
            v2_val = int(v2_parts[i]) if i < len(v2_parts) else 0

            if v1_val > v2_val:
                return 1
            elif v1_val < v2_val:
                return -1

        return 0


def main() -> None:
    sol = Solution165()
    v1 = "1.2"
    v2 = "1.10"
    print(sol.compareVersion(v1, v2))  # -1
    v1 = "1.01"
    v2 = "1.001"
    print(sol.compareVersion(v1, v2))  # 0
    v1 = "1.0"
    v2 = "1.0.0.0"
    print(sol.compareVersion(v1, v2))  # 0


if __name__ == "__main__":
    main()

File: compare-version-numbers/Solution165.java
// https://leetcode.com/problems/compare-version-numbers/

class Solution165 {

    // T: O(max(n,m)), S: O(n+m)
    public int compareVersion(String version1, String version2) {
        // Split versions into arrays of integers
        String[] v1Parts = version1.split("\\.");
        String[] v2Parts = version2.split("\\.");

        // Get the length of longer version number
        int maxLength = Math.max(v1Parts.length, v2Parts.length);

        // Compare each revision number
        for (int i = 0; i < maxLength; i++) {
            // Get current revision number, use 0 if index exceeds array length
            int v1Val = i < v1Parts.length ? Integer.parseInt(v1Parts[i]) : 0;
            int v2Val = i < v2Parts.length ? Integer.parseInt(v2Parts[i]) : 0;

            // Compare current revision numbers
            if (v1Val > v2Val) {
                return 1;
            } else if (v1Val < v2Val) {
                return -1;
            }
        }

        // All revision numbers are equal
        return 0;
    }

    public static void main(String[] args) {
        var sol = new Solution165();
        var v1 = "1.2";
        var v2 = "1.10";
        System.out.println(sol.compareVersion(v1, v2)); // -1
        v1 = "1.01";
        v2 = "1.001";
        System.out.println(sol.compareVersion(v1, v2)); // 0
        v1 = "1.0";
        v2 = "1.0.0.0";
        System.out.println(sol.compareVersion(v1, v2)); // 0
    }

}
